### ISSUE TITLE
Dask cluster hangs while processing file using s3fs

### ISSUE THREAD

**Issue by shalin07** [fsspec/s3fs#464]
So I have a function which loads nc files from S3 using s3fs and then select some specific region from files and then push back to different S3 bucket. Here's my function. I'm doing every thing in JupyterLab

```
def get_records(rec):
    
    d=[rec[-1][0:4], rec[-1][4:6], rec[-1][6:8], rec[-1][9:11], rec[-1][11:13]] 
    
    yr=d[0]
    mo=d[1]
    da=d[2]
    hr=d[3]
    mn=d[4]
    
    ps = s3fs.S3FileSystem(anon=True)

    period = pd.Period(str(yr)+str('-')+str(mo)+str('-')+str(da), freq='D')
    dy=period.dayofyear
    print(dy)

    cc=[7,8,9,10,11,12,13,14,15,16]  #look at the IR channels only for now
    dy="{0:0=3d}".format(dy)

    # this loop is for 10 different channels
    for c in range(10):
        ch="{0:0=2d}".format(cc[c])

    # opening 2 different time slices of given particular record
        F1=xr.open_dataset(ps.open(ps.glob('s3://noaa-goes16/ABI-L1b-RadF/'+str(yr)+'/'+str(dy)+'/'+str("{0:0=2d}".format(hr))+'/'+'OR_ABI-L1b-RadF-M3C'+ch+'*')[-2]))[['Rad']]

        F2=xr.open_dataset(ps.open(ps.glob('s3://noaa-goes16/ABI-L1b-RadF/'+str(yr)+'/'+str(dy)+'/'+str("{0:0=2d}".format(hr))+'/'+'OR_ABI-L1b-RadF-M3C'+ch+'*')[-1]))[['Rad']]
      
    # Selecting data as per given record radiance
        G1 = F1.where((F1.x >= (rec[0]-0.005)) & (F1.x <= (rec[0]+0.005)) & (F1.y >= (rec[1]-0.005)) & (F1.y <= (rec[1]+0.005)), drop=True)
        G2 = F2.where((F2.x >= (rec[0]-0.005)) & (F2.x <= (rec[0]+0.005)) & (F2.y >= (rec[1]-0.005)) & (F2.y <= (rec[1]+0.005)), drop=True)
       
    # Concating 2 time slices togethere
        G = xr.concat([G1, G2], dim  = 'time')

    # Concatiating different channels
        if c == 0:
            T = G    
        else:
            T = xr.concat([T, G], dim = 'channel')

    # Saving into nc file and storing them to S3
    path = rec[-1]+'.nc'
    T.to_netcdf(path)
    fs.put(path, bucket+path)
```
Now I want to use dask cluster to run this function parallelly like this this:

```
files = [ ]
for i in range(0, 100):
        s3_ds = dask.delayed(get_records)(records[i])
        files.append(s3_ds)

    files = dask.compute(*files)
``` 
So this thing was running perfectly fine last month but now when I try to do it again my clusters are not processing any files after a while. For example if I give 100 files to 10 cluster they won't process any files after 60-70 even though they have memory left and will just sit idle and do nothing. So next time I tried to give just 50 files, then they'll process around 30 files and then will sit idle. Even though they have memory left in them. I don't know what I'm doing wrong or is it some bug in library.  I upgraded all the library I was using, dask, s3fs and fssps. But still nothing is working. 

Main thing is all this was working perfectly fine few weeks back, but now same code is not working anymore.

**Environment**:

- Dask version: 2021.2.0
- Python version: 3.8.8
- S3fs version: 0.6.0
- Fsspec: 0.9.0

**Comment by jkornblum** [comment]
Its some issue with `fsspec==0.9.0` which was released April 5 2021.
You need to pin `fsspec` down to `0.8.7` and s3fs to `0.5.2`.
`fsspec==0.8.7` is not compatible with `s3fs==0.6.0` (import error) so you'll need to pin that down to `0.5.2`.

I think s3fs needs to change their `requirements.txt` for `0.6.x` to be `fsspec>=0.9.0`

This slipped into our production environment recently, it was a bit of a head scratcher.
Unit tests can't catch this, we need to work on some integration testing :)

**Comment by shalin07** [comment]
> Its some issue with `fsspec==0.9.0` which was released April 5 2021.
> You need to pin `fsspec` down to `0.8.7` and s3fs to `0.5.2`.
> `fsspec==0.8.7` is not compatible with `s3fs==0.6.0` (import error) so you'll need to pin that down to `0.5.2`.
> 
> I think s3fs needs to change their `requirements.txt` for `0.6.x` to be `fsspec>=0.9.0`
> 
> This slipped into our production environment recently, it was a bit of a head scratcher.
> Unit tests can't catch this, we need to work on some integration testing :)

Hello,

I also tried doing so earlier but still it didn't work me? :( 

Did that solve your problem?

**Comment by jkornblum** [comment]
Yes, this is what we've been running for a long while.
We pushed some updates recently and upgraded s3fs to `0.6.0` and this bug popped into production.
Pinning it back down as described resolved everything.

The behavior we saw is that basically any call to S3FileSystem just hung forever.

**Comment by shalin07** [comment]
> Yes, that (`s3fs==0.5.2` and `fsspec==0.8.x`) is what we've been running for a long while.
> We pushed some updates recently and upgraded s3fs to `0.6.0` and this bug popped into production.
> Pinning it back down as described resolved everything.
> 
> The behavior we saw is that basically any call to S3FileSystem just hung forever.

It doesn't seem to be working for me! :( what's the dask version you're using sir?

**Comment by jkornblum** [comment]
Oh sorry, this is outside of Dask.
We are using s3fs in a stand-alone Python application.

**Comment by martindurant** [comment]
@jkornblum , if you are seeing a hang outside of dask, that might be easier to debug. Please post your code, and turn on s3fs logging to see what it it doing at the time of hang. Can you interrupt the code and run a debugger?

**Comment by shalin07** [comment]
@martindurant hi, so I tried to debug my code line by line to see where it's getting stuck. So I added this line in my code for line by line debugging:

```
import ipdb; ipdb.set_trace()
```
```
for c in range(10): # c is channel
        ch="{0:0=2d}".format(cc[c])

    # opening 2 different time slices of given particular record
        F1=xr.open_dataset(ps.open(ps.glob('s3://noaa-goes16/ABI-L1b-RadF/'+str(yr)+'/'+str(dy)+'/'+str("{0:0=2d}".format(hr))+'/'+'OR_ABI-L1b-RadF-M3C'+ch+'*')[-2]))[['Rad']]

        F2=xr.open_dataset(ps.open(ps.glob('s3://noaa-goes16/ABI-L1b-RadF/'+str(yr)+'/'+str(dy)+'/'+str("{0:0=2d}".format(hr))+'/'+'OR_ABI-L1b-RadF-M3C'+ch+'*')[-1]))[['Rad']]
      
    # Selecting data as per given record radiance
        G1 = F1.where((F1.x >= (rec[0]-0.005)) & (F1.x <= (rec[0]+0.005)) & (F1.y >= (rec[1]-0.005)) & (F1.y <= (rec[1]+0.005)), drop=True)
        G2 = F2.where((F2.x >= (rec[0]-0.005)) & (F2.x <= (rec[0]+0.005)) & (F2.y >= (rec[1]-0.005)) & (F2.y <= (rec[1]+0.005)), drop=True)
       
    # Concating 2 time slices together
        G = xr.concat([G1, G2], dim  = 'time')

    # Concatiating different channels
        if c == 0:
            T = G    
        else:
            T = xr.concat([T, G], dim = 'channel')
```

And I found out that every time it's getting stuck in this part where it's pulling file from GOES:

```
 F1=xr.open_dataset(ps.open(ps.glob('s3://noaa-goes16/ABI-L1b-RadF/'+str(yr)+'/'+str(dy)+'/'+str("{0:0=2d}".format(hr))+'/'+'OR_ABI-L1b-RadF-M3C'+ch+'*')[-2]))[['Rad']]

        F2=xr.open_dataset(ps.open(ps.glob('s3://noaa-goes16/ABI-L1b-RadF/'+str(yr)+'/'+str(dy)+'/'+str("{0:0=2d}".format(hr))+'/'+'OR_ABI-L1b-RadF-M3C'+ch+'*')[-1]))[['Rad']]
      
```

And the most frustrating thing is it's getting stuck randomly like sometimes it'll get stuck while running the loop 2nd time which will be ```ch no 8``` which is basically one channel and next time when I run for the same thing it'll be able to pass that channel properly and then get stuck while running loop for 6th time which will be some other channel. So it's not some problem in code cause it's not getting stuck on one specific thing.

@martindurant can you tell me to try anything specific or decoding technique to make it more easier to catch! And rightnow I'm doing everything on my environment only not even on dask.

**Comment by swheaton** [comment]
I encountered the same problem, but I had s3fs=2021.4.0 and fsspec=2021.4.0. Not using dask or anything, just simply using s3fs to access files. My other packages could be playing into this relationship, but it is still pretty strange.

Some combinations I tried
s3fs=2021.4.0, fsspec=2021.4.0 -- NO, hangs
s3fs=0.6.0, fsspec=2021.4.0 -- NO, hangs
s3fs=0.5.2, fsspec=2021.4.0 -- NO, import error
s3fs=0.4.0, fsspec=2021.4.0 -- YES
s3fs=0.5.2, fsspec=0.8.7 -- YES
s3fs=0.6.0, fsspec=0.8.7 -- conda install incredibly unhappy (read: slow and found conflicts)
s3fs=0.6.0, fsspec=0.9.0 -- NO, hangs

**Comment by martindurant** [comment]
@swheaton , can you please give an idea of the operations you are running? This might just be the thing we can finally debug. It would be useful if you could set the environment variable S3FS_LOGGING_LEVEL=DEBUG too, so we can see what call(s) is being made. 

Also it may be worth trying with the current latest main version of s3fs, and/or with aiobotocore alone, if you have the time.

**Comment by swheaton** [comment]
Yes @martindurant, I was able to boil it down to the below code. It was a little more than 'simply using s3fs to access files' as I originally said, because we are using a multiprocess paradigm. I don't think it's the same issue as OP, because the code snippet does not hang using OP's python and s3fs/fsspec versions. Perhaps I should open under a new issue?

If broken, it will hang after printing 'created s3fs'. If not, it runs through and exits. In the hanging case, no debug output is given with S3FS_LOGGING_LEVEL.
```python
from multiprocessing import Process
import s3fs
def f():
    print('start')
    my_s3fs2=s3fs.S3FileSystem()
    print('created s3fs')
    f=my_s3fs2.open('s3://blah/blah', 'rb')
    print('opened')
if __name__ == '__main__':
    my_s3fs = s3fs.S3FileSystem()
    p = Process(target=f)
    p.start()
    p.join()
```
I also discovered it hangs in python 3.6 and 3.7, but not in >=3.8.
Attached:
1. [working_env.txt](https://github.com/dask/s3fs/files/6408631/working_env.txt) - conda env export, s3fs=2021.4.0, python=3.8.8
2. [hang_me.txt](https://github.com/dask/s3fs/files/6408630/hang_me.txt) - conda env export, s3fs=2021.4.0, python=3.7.10

**Comment by martindurant** [comment]
Can you please try with `multiprocessing.set_start_method('spawn')` (or 'forkserver') ?

**Comment by swheaton** [comment]
Thank you, using 'spawn' or 'forkserver' works, using 'fork' does not.

From the [docs](https://docs.python.org/3/library/multiprocessing.html)
> Note that safely forking a multithreaded process is problematic.

Is s3fs creating threads by merely instantiating s3fs.S3FileSystem? Line copied again below. If I remove the S3FileSystem instantiation, the hang occurs.
```python
if __name__ == '__main__':
    my_s3fs = s3fs.S3FileSystem()
    p = Process(target=f)
```

This would also explain why python 3.8 was working for me, as I was indeed running on Mac OS
> Changed in version 3.8: On macOS, the spawn start method is now the default. The fork start method should be considered unsafe as it can lead to crashes of the subprocess. See bpo-33725.

**Comment by martindurant** [comment]
Yes, creating an instance creates an event loop in a separate thread - both of which are likely to crash after `fork`.  The line [here](https://github.com/intake/filesystem_spec/blob/master/fsspec/asyn.py#L144) appears not to to catch the situation.

**Comment by swheaton** [comment]
That line would only fire if it were using the same s3fs instance inside the child process, but the child creates its own to use, so naturally the pids are equal. It could store to a class variable instead of / in addition to an instance variable?

So to be clear, are you saying that the code snippet above worked in earlier s3fs versions by "accident", and that it is the intention for this library to now explicitly disallow this by raising an exception?

**Comment by martindurant** [comment]
> That line would only fire if it were using the same s3fs instance inside the child process

Actually, instances are cached, so that if you use the same init kwargs, you get the same instance. In any case, the loop and thread are global and used by all instances unless you specify otherwise. Storing the PID elsewhere might not be a bad idea, though.

>  the code snippet above worked in earlier s3fs versions by "accident"

Previously, s3fs wasn't async, and didn't need to worry about loops and threads; and in the mean time, the default start-method changed.

> it is the intention for this library to now explicitly disallow this

There is probably no (easy) way to code around it, so raising seemed appropriate and was the intended behaviour.

**Comment by swheaton** [comment]
> Actually, instances are cached, so that if you use the same init kwargs, you get the same instance.

I see that to be true in single process code, however the cache is cleared [on this line](https://github.com/intake/filesystem_spec/blob/bdcdb304bea87c2d458c6045cfbf28d963fa715b/fsspec/spec.py#L64) when it determines it's inside a new process. My testing confirmed the two instances are unique by printing `id(my_s3fs)` and `id(my_s3fs2)`, as well as seeing `__init__` being reached.

> Previously, s3fs wasn't async, and didn't need to worry about loops and threads

So even if asynchronous=False for all created instances, it's still going to set up threads and event loops and such at a global level? Any way to avoid overcomplicating things in the straightforward case (as in pre-async versions of s3fs)?

**Comment by martindurant** [comment]
> however the cache is cleared on this line when it determines it's inside a new process

That's good! At least some fork protection is working. So we also need to clear the global event loop and thread in this case - but only once for all classes (so perhaps the parent AsyncFileSystem class does this).

> if asynchronous=False for all created instances, it's still going to set up threads and event loops and such at a global level

Yes

> if asynchronous=False for all created instances, it's still going to set up threads and event loops and such at a global level

I'm afraid not - the internal calls are all async. It would amount to writing two complete implementations, and loosing the async benefits of bulk operations.

**Comment by martindurant** [comment]
But really the short story is: don't use fork!

**Comment by normanrz** [comment]
> But really the short story is: don't use fork!

I think this would be a very important insight to be put in the documentation.
